dialogflow/stopwatch/functions/index.js

```diff
@@ -0,0 +1,118 @@
+/**
+* Copyright 2019 Google LLC
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+'use strict';
+
+const admin = require('firebase-admin');
+const functions = require('firebase-functions');
+const http = require('http');
+const https = require('https');
+
+admin.initializeApp(functions.config().firebase);
+var db = admin.firestore();
+
+function saveOAuthToken(context, oauthToken) {
+var docRef = db.collection('users').doc(context.auth.uid);
+
+var setToken = docRef.set({token: oauthToken});
+console.log("Received oauthToken", oauthToken);
+}
+
+function generateAccessToken(context, accessToken, tokenType) {
+// generate short lived access token
+// An object of options to indicate where to post to
+var post_options = {
+host: 'iamcredentials.googleapis.com',
+path: '/v1/projects/-/serviceAccounts/SERVICE-ACCOUNT-NAME@YOUR_PROJECT_ID.iam.gserviceaccount.com:generateAccessToken',
+method: 'POST',
+headers: {
+'Authorization': tokenType + ' ' + accessToken
+}
+};
+console.log('Got accessToken', accessToken);
+// Set up the request
+var oauthToken = '';
+var post_req = https.request(post_options, (res) => {
+console.log('Got OAuth Token');
+res.setEncoding('utf8');
+res.on('data', (chunk) => {
+oauthToken += chunk;
+});
+res.on('end', () => {
+// Next step in pipeline
+console.log("Calling save oauth token", oauthToken);
+saveOAuthToken(context, JSON.parse(oauthToken));
+})
+});
+
+post_req.on('error', (e) => {
+console.log('ERROR: ' + e.message);
+});
+
+const body = `{
+"delegates": [],
+"scope": [
+"https://www.googleapis.com/auth/dialogflow"
+],
+"lifetime": "3599s"
+}`;
+
+// post the data
+post_req.write(body);
+post_req.end();
+}
+
+
+exports.getToken = functions.https.onCall((data, context) => {
```

Comment:
Right now, this method generates a new token every time. 

Want want to make a few changes to this. 
1. We want to keep one global token used by all users in Firestore. 
2. Each time a user requests a token, the Firebase Function will check Firestore for the token. Validate it is not expired, then return that token. 
3. If the token is expired or will expire soon. We can generate a new token, save it to Firestore, then return the new token to the client.

Reply:
Yes, I have made the necessary changes to the index.js, Now we are maintaining one global token to be used by all users in Firestore. Also, fetching the generated token and expiry time.